wait for tcp port available (win32)

dmanto opened the pull request:
Wait for server ready on windows platforms. This helps to allow a minimal windows platforms support. Related to issue #1

kraih committed:
Link to blog post

kraih committed:
Use getters for pid and port

kraih committed:
Small documentation improvements

kraih committed:
Use a little less code

kraih committed:
The index.js file is not needed

kraih committed:
Bump version

kraih committed:
newServer returns a Promise

dmanto committed:
wait for tcp port available

kraih commented:
This module follows the normal Mojolicious rules. So tests and documentation are still required.

kraih commented:
Why is this a Windows only feature? I don't see any Windows specific code. It is actually counter to the whole argument for adding the feature in the first place. We could offer it for Windows, macOS and Linux., as a slightly less safe alternative to passing the file descriptor directly to the managed process.

kraih commented on lib/server-starter.js:
```diff
@@ -90,7 +92,23 @@ class Server extends EventEmitter {
     // Should be resolved by the "spawn" event, but that is only supported in Node 15+
-    return Promise.resolve;
+    if (process.platform !== 'win32') return Promise.resolve;
+    return new Promise((resolve) => {
```
There is no reject handler. What's supposed to happen when it times out?

dmanto replied:
the caller will do exactly the same (trying to connect), and will fail because the port is not ready.

kraih replied:
A promise does not resolve on its own.

kraih commented on lib/server-starter.js:
```diff
@@ -90,7 +92,23 @@ class Server extends EventEmitter {
+    if (process.platform !== 'win32') return Promise.resolve;
+    return new Promise((resolve) => {
+      const retryTime = 60;
+      const now = new Date();
```
This constant does nothing.

dmanto replied:
it is used in the following line to calculate timeToStop;

kraih replied:
But there is no use in giving it a name.

kraih replied:
`const timeToStop = new Date(Date.now() + win32Timeout);`?

dmanto replied:
Ok

kraih commented on lib/server-starter.js:
```diff
@@ -90,7 +92,23 @@ class Server extends EventEmitter {
-    return Promise.resolve;
+    if (process.platform !== 'win32') return Promise.resolve;
+    return new Promise((resolve) => {
+      const retryTime = 60;
```
Why hardcoded `60`?

dmanto replied:
most of the time with just one 60mS iteration (on my windows computer or in github workers) the port gets ready.
Anyway should be small so we don't add too much unnecesary time, but not that small because we don't want to block the thread

kraih replied:
So it should be a configurable interval.

dmanto replied:
Ok

kraih commented on lib/server-starter.js:
```diff
@@ -90,7 +92,23 @@ class Server extends EventEmitter {
+      (function loop () {
+        const connection = net.connect(port, resolve);
+        connection.on('error', err => {
+          if (err.code === 'ECONNREFUSED') {
```
And if it's something else?

dmanto replied:
should resolve the promise. But it actually doesn't, my mistake, see my answer about the logic a few minutes ago

kraih replied:
It should reject the promise, not resolve.

dmanto replied:
The only thing the loop does is to wait enough time to get the server in a defined (either right or wrong) state. So if there is an error different from ECONNREFUSED there is no need to stay waiting, there is no argue on that.

But I think there is no added value in reject the promise with an error either. This is because you will try to connect by yourself after the launch of the server, and you will be sucesfull or not, and get the right error in any case.

Anyway the promise could be rejected when there is an error detected, the change itself is trivial (although the specific tests are not that easy) but I cannot think of any use case in which this can be beneficial for the user of the module.

kraih commented on lib/server-starter.js:
```diff
@@ -90,7 +92,23 @@ class Server extends EventEmitter {
+        const connection = net.connect(port, resolve);
+        connection.on('error', err => {
+          if (err.code === 'ECONNREFUSED') {
+            if (new Date() < timeToStop) {
```
And if it times out we just do nothing?

dmanto replied:
yes the caller will take care, I mean will recive the exact same error

kraih replied:
How does the error reach the caller?

dmanto replied:
The logic is like:

- wait up to win32Timeout mS, as long as the error you are getting is ECONNREFUSED.
- if you could connect, resolve the promise
- if you get other error, you will not retry

so Ooops, I need to resolve at the end of the anonymous function (or reject, but that will be handled much better by the caller, when it tries to connect)

kraih commented on lib/server-starter.js:
```diff
@@ -64,12 +64,14 @@ class Server extends EventEmitter {
    * @param {object} [options] - Optional settings
    * @param {boolean} [options.stdout=false] - Forward server output from STDOUT to STDOUT
    * @param {boolean} [options.stderr=true] - Forward server output from STDERR to STDERR
+   * @param {number} [options.win32Timeout=30000] - (Windows only) Max time to wait for server ready, in mS
```
As far as i can see there is no reason at all for this to be Windows specific.

dmanto replied:
It will have no effect on other platforms, at least how is it written.

If you mean to wait for connection ready at the TCP port in all cases, it could be done.

kraih replied:
The way it is done now makes little sense. Since the module doesn't become more portable, it just allows someone to use it in a Windows specific way. Instead of enabling one test to run on all platforms.

kraih commented:
So, to sum up the requirements. a) The feature needs to work on all platforms, Windows, macOS and Linux. b) There needs to be an additional documentation section explaining how to use the module in a portable (but slightly less safe) way. c) There need to be new tests for success (random and specific port) and failure cases (timeout, failed web app start, connection error other than ECONNREFUSED).

kraih committed:
Enable mergify

kraih committed:
Make room for port based tests

kraih committed:
Test port assignment

kraih committed:
Give the test a little more time to close the port

kraih committed:
Wait for the first listen socket to be closed

dmanto committed:
add listenAddress attribute

dmanto committed:
avoid fdpass through env var

dmanto committed:
fix lint errors

dmanto committed:
add slow server tests

dmanto committed:
fix lint again

dmanto committed:
modify README for non fd passing cases

dmanto committed:
add minimal windows support

dmanto committed:
small fixes

dmanto committed:
add avoidFdPassing option to launch funtion

dmanto committed:
remove listenAddress() function, improve README

dmanto committed:
add race condition caveat

dmanto committed:
fix typos and rephrase race condition caveat

dmanto committed:
fix typos

dmanto committed:
add windows workflow, but skiping start_fd.js tests in this case

dmanto committed:
reorganize README

dmanto committed:
manual merge into former branch

dmanto committed:
fix workflow

dmanto committed:
leave server_fd.js as in master branch

dmanto committed:
fix typo

dmanto committed:
add a launchPortable() function instead of checking avoidFdPassing option, as logic turned out to be very different

dmanto committed:
fix README intralink

dmanto committed:
more tests on portable mode

dmanto commented:
- I hope we still have a chance to get to a usable feature on this issue.
- And as the other PR was a mistake, so we leave it closed, but I will address all issues mentioned by you in the following points.
- Regarding your sum up on requirements above:
     a) The feature should work in Linux, macOS and Windows: done, and tested on all of them,
     b) Additional documentation: done, added a section on launching servers with this "portable" feature,
     c) New tests: (all of them in a new script:  test/start_port.js, which is a modified version of test/start_fd.js)
       - success (random port): done,
       - success (specific port): done,
       - timeout, done (reduced to 500 mS so we don't slow down tests too much),
       - failed web app start: done, we got a timeout plus server emitted specific error,
       - connection error other than ECONNREFUSED: done, we got a timeout plus server emitted specific error
       Also regarding to tests, I needed a way to skip fd tests on windows, so I used some tap functionality to skip them all on win32 platforms. It may look hacky, so just let me know if you know a cleaner way.
- About the "no reject handler" remark: the reject case was added so is fixed now (done). Also I want to apologize for my misunderstanding on the process, I answered in the conversation with some explanation in the timeout logic and thought that was enough, that's why I included the comment in the code (In no way I meant to offend you nor to ignore any of your comments).
- About the "this constant does nothing" remark, sorry I had missed your last [comment](https://github.com/mojolicious/server-starter/pull/2#discussion_r605083757), is fixed now (done).
- About the "me still missing the point" as this is not meant to be a Mojolicious specific module, I think I understand that. I removed the ENV variable used to avoid fd passing, and also the "listenAddress()" function, so to avoid any "Hidden magic" as you mentioned on other remark.
- To let the customer choose how to launch the child server process, a new launch function, launchPortable(), is defined.
The reason I separate both cases is that the logic changes a lot between them, as launchPortable needs the port to be closed before spawning the server, otherwise the "first" race condition will trigger a lot (almost always on a mac). launch() in the other hand, needs to close the port after the server opened the passed fd. And for the retry and timeout functionality, I cannot see any case in which that functionality would make sense when using fd passing, as the port should be up all the time.

dmanto committed:
fix node 10.x windows cannot find module error

dmanto committed:
fix bad port test on windows node 10.x

dmanto committed:
fix bad port test, try harder

dmanto committed:
skip some tests on node 10.x (windows), seems to have error msgs broken
